NIFI-1582 added state to UpdateAttribute as well as updated a few par…

JPercivall opened the pull request:
…ts that hadn't be touched in years (referenced the 'FlowFileMetadataEnhancer' processor'. Also added a 'NUMBER_VALIDATOR' to StandardValidators

JPercivall commented:
Not sure why it failed on something unrelated on Travis it builds locally. Also I don't believe there is an easy way to rebuild the PR in Travis.

mattyb149 commented:
I think you can re-push the PR (no new commits) and it might pick that up and rebuild.

JPercivall commented:
@mattyb149 just tried pushing with no new commits and nothing happened. I believe I could push an empty commit but I don't think it's worth cluttering this PR with unnecessary commits.

mattyb149 commented:
There was something that could be done, I can't remember. Maybe reset --soft HEAD^ and re-commit and push -f. Will give the same code a new commit SHA and should kick it off again. I thought it was simpler to do but maybe not. Also maybe we can re-run the Travis job manually?

apiri commented:
There are no facilities to manually start a Travis job.  You could likely squash your two commits and push those.  The core thing is that a new hash is seen for the Travis webhooks, I believe.

JPercivall commented:
@mattyb149 I did what you suggested, did a soft reset, re-committed (making a new hash like @apiri pointed out) and did a force push. It's rebuilding now

olegz commented on nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java:
```diff
@@ -187,24 +167,83 @@ public UpdateAttribute() {
+                    .expressionLanguageSupported(true)
+                    .dynamic(true)
+                    .build();
+        }
```
Joe, instead of repeating the entire definition of the descriptor, have you considered a base builder and then finish descriptor based on condition. For example
```
Builder somePropBuilder = new PropertyDescriptor.Builder()
            .name(propertyDescriptorName)
            .required(false)
            .addValidator(StandardValidators.ATTRIBUTE_KEY_PROPERTY_NAME_VALIDATOR)
            .expressionLanguageSupported(true)
            .dynamic(true);

if(!stateful){
        return somePropBuilder
               .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
               .build();
} else {
       return somePropBuilder
               .addValidator(StandardValidators.createAttributeExpressionLanguageValidator(AttributeExpression.ResultType.STRING, true))
               .build();
}
```

JPercivall replied:
That would be a cleaner approach. I will change. Thanks!

olegz commented on nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java:
```diff
@@ -187,24 +167,83 @@ public UpdateAttribute() {
+            } else {
+                stateful = false;
+            }
+        }
```
Isn't it the same as saying 
```
stateful = "true".equalsIgnoreCase(newValue);
```

JPercivall replied:
Yup, will fix

olegz commented on nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java:
```diff
@@ -489,17 +546,32 @@ private FlowFile executeActions(final ProcessSession session, final ProcessConte
+
         // go through each action
         for (final Action action : actions.values()) {
             if (!action.getAttribute().equals(DELETE_ATTRIBUTES.getName())) {
```
Any possibility of getAttribute() returning null? Just want to avoid unnecessary NPE.

JPercivall replied:
I didn't write this piece of code so I'm not entirely sure but I highly doubt it since its been on master for a while and hasn't been a problem so far.

olegz commented on nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java:
```diff
@@ -489,17 +546,32 @@ private FlowFile executeActions(final ProcessSession session, final ProcessConte
                     }
 
+                    if (statefulAttributesToSet != null) {
+                        if(!action.getAttribute().equals("UpdateAttribute.matchedRule")) {
```
Same comment as above on NPE

trixpan commented:
@JPercivall welcome back!

Could you rebase this PR?

JPercivall commented:
Thanks @trixpan! Rgr, I'm working on rebasing this and making the state variable accessing a bit more user-friendly

trixpan commented:
@JPercivall no issue. let us know when ready to review

JPercivall commented:
@trixpan, Ah sorry for not commenting but it is ready for review. It is rebased and I added an EL function to access state in a consistent user-friendly way.

JPercivall commented:
Hey @trixpan, if you have free cycles I'd appreciate finalizing this review. I've rebased on the latest master

JPercivall commented on nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java:
```diff
@@ -162,20 +146,33 @@ public ValidationResult validate(String subject, String input, ValidationContext
+            .description("If using state to set/reference variables then this value is used to set the initial value of the stateful variable. This will only be used in the @OnScheduled method " +
+                    "when state does not contain a value for the variable.")
+            .required(false)
+            .defaultValue("0")
```
Yeah I agree that it should be empty. I will change

JPercivall commented:
@mattyb149 removed the default value, added a validation check so that it is invalid if it is configured for state but has no initial value, and added a check in the unit tests.

mattyb149 commented on nifi-nar-bundles/nifi-update-attribute-bundle/nifi-update-attribute-processor/src/main/java/org/apache/nifi/processors/attributes/UpdateAttribute.java:
```diff
@@ -166,7 +166,6 @@ public ValidationResult validate(String subject, String input, ValidationContext
                     "when state does not contain a value for the variable.")
             .required(false)
-            .defaultValue("0")
             .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
If we want to allow empty values, we shouldn't have the NON_EMPTY_VALIDATOR here. Also the description should mention that the user must Set Empty String if they want the value to exist but be empty. That doesn't change behavior per se, but not setting this property (versus setting it to empty) will affect the additional check for null property value below

JPercivall commented:
@mattyb149 pushed out a new commit adding documentation on the init value requirements and clarification on the "empty" value. 

Thanks for reviewing documentation too! I totally forgot it.

mattyb149 commented:
+1 LGTM, thanks for this awesome feature! I ran the tests and tried with various incrementing strategies, all looks good, merging to master
